packages/code-infra/src/cli/cmdPublishCanary.mjs

```diff
@@ -8,26 +8,383 @@
+        return acc;
+      }
+      const deps = Object.keys(pkg.dependencies || {});
+      if (!deps.length) {
+        return acc;
+      }
+      acc[pkg.name] = deps;
+      return acc;
+    }, /** @type {Record<string, string[]>} */ ({}));
+
+  // Compute transitive (nested) dependencies limited to workspace packages and avoid cycles.
+  const workspacePkgNames = new Set(Object.keys(directPkgDependencies));
+  const nestedMap = /** @type {Record<string, string[]>} */ ({});
+
+  /**
+   *
+   * @param {string} pkgName
+   * @returns {string[]}
+   */
+  const getTransitiveDeps = (pkgName) => {
+    /**
+     * @type {Set<string>}
+     */
+    const seen = new Set();
+    const stack = (directPkgDependencies[pkgName] || []).slice();
+
+    while (stack.length) {
+      const dep = stack.pop();
+      if (!dep || seen.has(dep)) {
+        continue;
+      }
+      // Only consider workspace packages for transitive expansion
+      if (!workspacePkgNames.has(dep)) {
+        // still record external deps as direct deps but don't traverse into them
+        seen.add(dep);
+        continue;
+      }
+      seen.add(dep);
+      const children = directPkgDependencies[dep] || [];
+      for (const c of children) {
+        if (!seen.has(c)) {
+          stack.push(c);
+        }
+      }
+    }
+
+    return Array.from(seen);
+  };
+
+  for (const name of Object.keys(directPkgDependencies)) {
+    nestedMap[name] = getTransitiveDeps(name);
+  }
+
+  return nestedMap;
+}
+
+/**
+ * Prepare changelog data for packages using GitHub API
+ * @param {PublicPackage[]} packagesToPublish - Packages that will be published
+ * @param {PublicPackage[]} allPackages - All packages in the repository
+ * @param {Map<string, string>} canaryVersions - Map of package names to their canary versions
+ * @returns {Promise<Map<string, string[]>>} Map of package names to their changelogs
+ */
+async function prepareChangelogsFromGitCli(packagesToPublish, allPackages, canaryVersions) {
+  /**
+   * @type {Map<string, string[]>}
+   */
+  const changelogs = new Map();
+
+  await Promise.all(
+    packagesToPublish.map(async (pkg) => {
+      const commits = await fetchCommitsForPackage({ packagePath: pkg.path });
+      if (commits.length > 0) {
+        console.log(`Found ${commits.length} commits for package ${pkg.name}`);
+      }
+      const changeLogStrs = commits
+        // Exclude commits authored by bots
+        .filter(
+          // We want to allow commits from copilot or other AI tools, so only filter known bots
+          (commit) => !AUTHOR_EXCLUDE_LIST.includes(commit.author),
+        )
+        .map((commit) => `- ${cleanupCommitMessage(commit.message)} by ${commit.author}`);
+
+      if (changeLogStrs.length > 0) {
+        changelogs.set(pkg.name, changeLogStrs);
+      }
+    }),
+  );
+  // Second pass: check for dependency updates in other packages not part of git history
+  const pkgDependencies = await getPackagesWithDependencies();
```

Comment:
nit: the flow would be clearer if this was composed of two steps. Now the naming suggests it returns a dependency map, but it's actually deeply resolved transitive dependencies

```tsx
const pkgDependencies = await getPackagesWithDependencies();
const transitiveDependencies = resolveTransitiveDependencies(pkgDependencies)
```